restore: Add support for sparse files

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
It adds a `restore --sparse` option to restore files as sparse. The detection of sparseness is done at restore time, by replacing long stretches of zeros by sparse sections.

Note that this does not mean that we are no longer interested in having the archiver store precise information about which files have gaps at which places. Instead this is intended to provide basic support for sparse files which can be improved upon later.  In particular, if sparseness information was stored for a file, then it would also an option to have the restorer automatically restore such files as sparse.

The implementation is largely based on #2601, with a few modifications. The `--sparse` option was initially suggested by #2378. The restorer preallocates files by default which conflicts with restoring files as sparse. The currently taken approach is to check whether a file contains an all zeros blocks and mark is as sparse if the option was set. This then disables preallocation for that specific file. That way most files can still be preallocated. The downside is that we cannot detect even smaller sparse sections, however, these were only partially handled before.

Another change is that a sparse file is now immediately truncated to it's final size. `partialFile.WriteAt` may be called in parallel by the restorer, which could result in race conditions between the `Truncate` and `WriteAt` calls.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Supersedes #2378
Supersedes #2601
Fixes #79

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [ ] I'm done! This pull request is ready for review.

Commit:
Write sparse files in restorer

This writes files by using (*os.File).Truncate, which resolves to the
truncate system call on Unix.

Compared to the naive loop,

	for _, b := range p {
		if b != 0 {
			return false
		}
	}

the optimized allZero is about 10× faster:

name       old time/op    new time/op     delta
AllZero-8    1.09ms ± 1%     0.09ms ± 1%    -92.10%  (p=0.000 n=10+10)

name       old speed      new speed       delta
AllZero-8  3.84GB/s ± 1%  48.59GB/s ± 1%  +1166.51%  (p=0.000 n=10+10)

Commit:
Changelog entry for sparse file restoring

MichaelEischer committed:
Enable sparseness only conditionally

We can either preallocate storage for a file or sparsify it. This
detects a pack file as sparse if it contains an all zero block or
consists of only one block. As the file sparsification is just an
approximation, hide it behind a `--sparse` parameter.

MichaelEischer committed:
restorer: Fix race condition in partialFile.WriteAt

The restorer can issue multiple calls to WriteAt in parallel. This can
result in unexpected orderings of the Truncate and WriteAt calls and
sometimes too short restored files.

DRON-666 commented on internal/restorer/fileswriter.go:
```diff
@@ -53,39 +57,45 @@ func (w *filesWriter) writeToFile(path string, blob []byte, offset int64, create
-				// This should yield a syscall.ENOTSUP error, but some other errors might also
-				// show up.
-				debug.Log("Failed to preallocate %v with size %v: %v", path, createSize, err)
+			if sparse {
```
You can also support Windows with just two lines of code (in separate file with `//go:build windows`, of course):
```
var t uint32
_ = windows.DeviceIoControl(windows.Handle(f.Fd()), windows.FSCTL_SET_SPARSE, nil, 0, nil, 0, &t, nil)
```

MichaelEischer committed:
restorer: move zeroPrefixLen to restic package

MichaelEischer committed:
repository: special case SaveBlob for all zero chunks

Sparse files contain large regions containing only zero bytes. Checking
that a blob only contains zeros is possible with over 100GB/s for modern
x86 CPUs. Calculating sha256 hashes is only possible with 500MB/s (or
2GB/s using hardware acceleration). Thus we can speed up the hash
calculation for all zero blobs (which always have length
chunker.MinSize) by checking for zero bytes and then using the
precomputed hash.

The all zeros check is only performed for blobs with the minimal chunk
size, and thus should add no overhead most of the time. For chunks which
are not all zero but have the minimal chunks size, the overhead will be
below 2% based on the above performance numbers.

This allows reading sparse sections of files as fast as the kernel can
return data to us. On my system using BTRFS this resulted in about
4GB/s.

MichaelEischer committed:
update sparse restore changelog

MichaelEischer committed:
restore: support sparse restores also on windows

MichaelEischer commented:
I've added the optimization discussed in #3903 to speed up hashing of all-zero chunks. On windows, files are now also marked as sparse.

MichaelEischer commented:
LGTM. I've rebased the PR and fix the returned byte count of `partialFile.WriteAt` for a sparse write.

FATruden commented:
How about add `--sparse` option to `dump` command? I think It will be very useful especially when we speak about restore of VM disk images which is usually have a lot of sparse sections.

FATruden commented:
Oh, I realized that dump with `--sparse` is not possible, because restic just stream bytes to stdout and don't do write directly to file on filesystem. Hm.. But may be It possible don't read all zeros from repo and just "generate" them on the fly)
But it will require to store information about zeros in some place...
